File: qssort.py
from sys import argv
from collections import deque

#open and parse file
with open(argv[1]) as f:
    M=int(f.readline())
    queue=[int(x) for x in (f.readline()).split()]

solved=False

#initial state is our input queue, empty stack, and empty moves
init=tuple(queue),(),""
#sorting queue to compare later
final=tuple(sorted(queue))

#set that keeps track of states we have visited
#only keep queue and stack
prev={(init[0],init[1])}

#yields states that state s can change into with one move
def next(s):
    #booleans to check what move we can do
    qq,ss= False, False

    #Q move
    if s[0]!=():  #check if queue is empty
        i=1
        for x in s[0][:-1]:
            if s[0][i]==s[0][i-1]:
                i+=1
            else: break
        qqueue=(s[0])[i:]
        qstack=(s[0])[:i]+s[1]
        qmoves=s[2]+i*"Q"
        qq=True

    #S move
    if s[1]!=():  #check if stack is empty
        i=1
        for x in s[1][:-1]:
            if s[1][i]==s[1][i-1]:
                i+=1
            else: break
        sstack=(s[1])[i:]
        squeue=s[0]+(s[1])[:i]
        smoves=s[2]+i*"S"
        ss=True

    if qq and ss:
        yield (qqueue,qstack,qmoves),(squeue,sstack,smoves)
    elif qq:
        yield (qqueue,qstack,qmoves),init
    else:
        yield (squeue,sstack,smoves),init


#deque of states we visit
Q=deque([init])

while Q:
    s=Q.popleft()
    if s[0]==final:
        solved=True
        break
    for t in next(s):
        for a in t:
            if (a[0],a[1]) not in prev:
                Q.append(a)
                prev.add((a[0],a[1]))

if solved:
    if s[2]=="":
        print('empty')
    else:
        print(s[2])

File: qssort.pl
/*
 * A predicate that reads the input from File and returns it in
 * the last three arguments: N, K and C.
 * from course website
 */
read_input(File, M, D) :-
    open(File, read, Stream),
    read_line(Stream, M),
    read_line(Stream, D).

read_line(Stream, L) :-
    read_line_to_codes(Stream, Line),
    atom_codes(Atom, Line),
    atomic_list_concat(Atoms, ' ', Atom),
    maplist(atom_number, Atoms, L).



% do a Q move
% if same elements, move them together
qmove([H|T],S,M,Q2,S2):-
  (T=[H|_] -> append([q],M2,M), qmove(T,[H|S],M2,Q2,S2)
  ;Q2=T, S2=[H|S], M=[q]).

% do an S move
% if same elements, move them together
smove(Q,[H|T],M,Q2,S2):-
  (T=[H|_] -> append([s],M2,M), append(Q,[H],Q3), smove(Q3,T,M2,Q2,S2)
  ;append(Q,[H],Q2), S2=T, M=[s]).

% Check for Q move before S move
move([H|T],S,M,Q2,S2):-
  qmove([H|T],S,M,Q2,S2).
move(Q,[H|T],M,Q2,S2):-
  smove(Q,[H|T],M,Q2,S2).

% predicate where Moves is a list of q,s moves that sort Q into Sorted
% returns results in order of least amount of moves by calling length
% calls helper predicate with 4 arguments, starting with empty stack
solve(Q,Moves,Sorted):-
  length(Moves,_),
  solve(Q,[],Moves,Sorted).
% helper predicate
% solved if stack is empty and queue Q is sorted
solve(Q,[],_,Sorted):-
  Q=Sorted.
% helper predicate
% Q is current queue, S is current stack, Sorted is final result
% Q2 and S2 are queue and stack after Move
% Moves is list of moves
solve(Q,S,[Move|Moves],Sorted):-
  move(Q,S,Move,Q2,S2),
  solve(Q2,S2,Moves,Sorted).
% to give only one solution
solve1(Q,Moves,Sorted):-
  once(solve(Q,Moves,Sorted)).


qssort(File,Answer) :-
  read_input(File,_,D),
  msort(D,Sorted),
  (Sorted=D -> Answer="empty"
  ; solve1(D,Moves,Sorted),
    flatten(Moves,Flat),
    atomics_to_string(Flat,Lower),
    string_upper(Lower,Answer)
    ).
